File: code/JL_OTA_V1.7.1_SDK_V1.9.2/otasdk/src/main/java/com/jieli/otasdk/dialog/DialogOTA.java
package com.jieli.otasdk.dialog;

import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;

import com.jieli.jl_bt_ota.constant.ErrorCode;
import com.jieli.jl_bt_ota.constant.JL_Constant;
import com.jieli.jl_bt_ota.util.JL_Log;
import com.jieli.otasdk.R;
import com.jieli.otasdk.base.BaseDialogFragment;
import com.jieli.otasdk.databinding.FragmentDialogOtaBinding;
import com.jieli.otasdk.model.ota.OTAEnd;
import com.jieli.otasdk.model.ota.OTAState;
import com.jieli.otasdk.model.ota.OTAWorking;
import com.jieli.otasdk.tool.config.ConfigHelper;
import com.jieli.otasdk.viewmodel.OTAViewModel;

import java.util.Locale;

/**
 * OTA进度提示
 */
public class DialogOTA extends BaseDialogFragment {
    private int testOTAType = ConfigHelper.Companion.getInstance().isAutoTest() ? 1 : 0;//0：简单的单次测试 1：自动化测试OTA
    private FragmentDialogOtaBinding mBinding;
    private OTAViewModel otaViewModel = null;
    private String TAG = this.getClass().getSimpleName();
    private boolean isOTAError = false;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        //设置dialog的基本样式参数
        requireDialog().requestWindowFeature(Window.FEATURE_NO_TITLE);
        Window window = requireDialog().getWindow();
        if (window != null) {
            //去掉dialog默认的padding
            window.getDecorView().setPadding(0, 0, 0, 0);
            WindowManager.LayoutParams layoutParams = window.getAttributes();
            layoutParams.width = WindowManager.LayoutParams.MATCH_PARENT;
            layoutParams.height = WindowManager.LayoutParams.WRAP_CONTENT;
            layoutParams.gravity = Gravity.BOTTOM;
            window.setAttributes(layoutParams);
            window.setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        }
        mBinding = FragmentDialogOtaBinding.inflate(inflater, container, false);
        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        otaViewModel = new ViewModelProvider(requireActivity()).get(OTAViewModel.class);
        if (testOTAType == 0) {
            initUISingleTest();
        } else {
            initUIAutoTest();
        }
        observeCallback();
        mBinding.tvSureBtn.setOnClickListener(v -> {
            otaViewModel.getOtaStateMLD().setValue(null);
            dismiss();
        });
    }

    //单次测试
    private void initUISingleTest() {
        mBinding.groupAutoTestTitle.setVisibility(View.GONE);
        FrameLayout.LayoutParams layoutParams = (FrameLayout.LayoutParams) mBinding.clOtaContent.getLayoutParams();
        layoutParams.height = getPixelsFromDp(148);
    }

    //自动化测试OTA
    private void initUIAutoTest() {
        mBinding.groupAutoTestTitle.setVisibility(View.VISIBLE);
        FrameLayout.LayoutParams layoutParams = (FrameLayout.LayoutParams) mBinding.clOtaContent.getLayoutParams();
        layoutParams.height = getPixelsFromDp(223);
    }

    //增加观察者
    private void observeCallback() {
        otaViewModel.getOtaStateMLD().observe(getViewLifecycleOwner(), otaState -> {
            if (otaState == null) return;
            JL_Log.d(TAG, "otaStateMLD : >>>" + otaState);
            switch (otaState.getState()) {
                case OTAState.OTA_STATE_START://OTA开始
                    requireActivity().getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
                    mBinding.groupUpgrade.setVisibility(View.VISIBLE);
                    mBinding.groupScanDeviceLoading.setVisibility(View.GONE);
                    mBinding.tvUpgradeProgress.setText(getString(R.string.ota_check_file) + "  " + "0%");
                    mBinding.pbUpgradeProgress.setProgress(0);
                    break;
                case OTAState.OTA_STATE_RECONNECT:
                    mBinding.groupUpgrade.setVisibility(View.GONE);
                    mBinding.groupScanDeviceLoading.setVisibility(View.VISIBLE);
                    mBinding.tvOtaFileName.setVisibility(View.GONE);
                    mBinding.tvUpgradeProgress.setText(getString(R.string.ota_upgrading) + "  " + "0%");
                    mBinding.pbUpgradeProgress.setProgress(0);
                    mBinding.tvScanDeviceLoading.setText(R.string.verification_file_completed);
                    break;
                case OTAState.OTA_STATE_WORKING://OTA进行中
                    OTAWorking otaWorking = (OTAWorking) otaState;
                    String message = otaWorking.getType() == JL_Constant.TYPE_CHECK_FILE ? getString(R.string.ota_check_file) : getString(R.string.ota_upgrading);
                    int progress = Math.round(otaWorking.getProgress());
                    mBinding.groupUpgrade.setVisibility(View.VISIBLE);
                    mBinding.groupScanDeviceLoading.setVisibility(View.GONE);
                    if (testOTAType == 1) {
                        mBinding.tvOtaFileName.setVisibility(View.VISIBLE);
                    } else {
                        mBinding.tvOtaFileName.setVisibility(View.GONE);
                    }
                    mBinding.tvUpgradeProgress.setText(message + "  " + progress + "%");
                    mBinding.pbUpgradeProgress.setProgress(progress);
                    break;
                case OTAState.OTA_STATE_IDLE://OTA结束
                    requireActivity().getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
                    boolean isUpgradeSuccess = false;
                    OTAEnd otaEnd = (OTAEnd) otaState;
                    switch (otaEnd.getCode()) {
                        case ErrorCode.ERR_NONE://升级完成
                            mBinding.tvUpgradeProgress.setText(getString(R.string.ota_upgrading) + "  " + 100 + "%");
                            mBinding.pbUpgradeProgress.setProgress(100);
                            // TODO: 2022/11/1  单词升级结束
                            isUpgradeSuccess = true;
                            break;
                        case ErrorCode.ERR_UNKNOWN://升级取消
                            break;
                        default://升级失败
                            if (otaEnd.getCode() == ErrorCode.SUB_ERR_OTA_IN_HANDLE) {//正在升级中
                                return;
                            } else if (otaEnd.getCode() == ErrorCode.SUB_ERR_DATA_NOT_FOUND) {//未找到升级文件
                                otaViewModel.readFileList();
                            }
                            String otaMsg = String.format(Locale.getDefault(), "code:%d, %s", otaEnd.getCode(), otaEnd.getMessage());
                            mBinding.tvUpgradeResultReason.setText(getString(R.string.ota_reason) + otaMsg);
                            isOTAError = true;
                            break;
                    }
                    if (testOTAType == 0) {//单次升级
                        mBinding.groupScanDeviceLoading.setVisibility(View.GONE);
                        mBinding.groupUpgrade.setVisibility(View.GONE);
                        mBinding.groupUpgradeResult.setVisibility(View.VISIBLE);
                        FrameLayout.LayoutParams layoutParams = (FrameLayout.LayoutParams) mBinding.clOtaContent.getLayoutParams();
                        if (isUpgradeSuccess) {
                            mBinding.ivUpgradeResultLogo.setImageResource(R.drawable.ic_success_big);
                            mBinding.tvUpgradeResultTip.setText(R.string.update_finish);
                            layoutParams.height = getPixelsFromDp(196);
                        } else {
                            mBinding.ivUpgradeResultLogo.setImageResource(R.drawable.ic_fail_big);
                            mBinding.tvUpgradeResultTip.setText(R.string.update_failed);
                            mBinding.tvUpgradeResultReason.setVisibility(View.VISIBLE);
                            layoutParams.height = FrameLayout.LayoutParams.WRAP_CONTENT;
                        }
                        mBinding.clOtaContent.setLayoutParams(layoutParams);
                    }
                    break;

            }
        });
    }

    private int getPixelsFromDp(int size) {
        DisplayMetrics metrics = new DisplayMetrics();
        this.getActivity().getWindowManager().getDefaultDisplay().getMetrics(metrics);
        return (size * metrics.densityDpi) / DisplayMetrics.DENSITY_DEFAULT;
    }
}

File: code/JL_OTA_Android_V1.9.0_SDK_V1.11.0/otasdk/src/main/java/com/jieli/otasdk/ui/dialog/DialogOTA.kt
package com.jieli.otasdk.ui.dialog

import android.os.Bundle
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.FrameLayout
import com.jieli.jl_bt_ota.constant.ErrorCode
import com.jieli.jl_bt_ota.constant.JL_Constant
import com.jieli.jl_bt_ota.util.JL_Log
import com.jieli.otasdk.R
import com.jieli.otasdk.data.constant.OtaConstant
import com.jieli.otasdk.data.model.ota.OTAEnd
import com.jieli.otasdk.data.model.ota.OTAState
import com.jieli.otasdk.data.model.ota.OTAWorking
import com.jieli.otasdk.databinding.FragmentDialogOtaBinding
import com.jieli.otasdk.ui.ota.OTAViewModel
import com.jieli.otasdk.util.gone
import com.jieli.otasdk.util.show

/**
 * OTA进度提示
 */
class DialogOTA private constructor(builder: Builder) : CommonDialog(builder) {
    private lateinit var binding: FragmentDialogOtaBinding
    private var isOTAError = false

    override fun createView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        FragmentDialogOtaBinding.inflate(inflater, container, false).also {
            binding = it
            return it.root
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initUI()
        addObserver()
    }

    private fun initUI() {
        if (builder !is Builder) return
        val isAutoTest = builder.viewModel.isAutoTestOTA()
        binding.groupAutoTestTitle.apply {
            if (isAutoTest) show() else gone()
            binding.clOtaContent.let {
                it.layoutParams.let { lp ->
                    lp.height = getPixelsFromDp(if (isAutoTest) 223 else 148)
                    it.layoutParams = lp
                }
            }
        }
        binding.tvSureBtn.setOnClickListener {
            builder.viewModel.otaStateMLD.value = null
            dismiss()
        }
    }

    //增加观察者
    private fun addObserver() {
        if (builder !is Builder) return
        builder.viewModel.otaStateMLD.observe(viewLifecycleOwner) { otaState ->
            if (otaState == null) return@observe
            handleOtaState(otaState)
        }
    }

    private fun handleOtaState(otaState: OTAState) {
        if (builder !is Builder || !isAdded) return
        JL_Log.d(TAG, "handleOtaState", otaState.toString())
        val isAutoTest = builder.viewModel.isAutoTestOTA()
        when (otaState.state) {
            OTAState.OTA_STATE_START -> {
                requireActivity().window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
                binding.groupUpgrade.show()
                binding.groupScanDeviceLoading.gone()
                binding.tvUpgradeProgress.text =
                    formatProgressText(getString(R.string.ota_check_file), 0)
                binding.pbUpgradeProgress.progress = 0
            }

            OTAState.OTA_STATE_RECONNECT -> {
                binding.groupUpgrade.gone()
                binding.groupScanDeviceLoading.show()
                binding.tvOtaFileName.gone()
                binding.tvUpgradeProgress.text =
                    formatProgressText(getString(R.string.ota_upgrading), 0)
                binding.pbUpgradeProgress.progress = 0
                binding.tvScanDeviceLoading.setText(R.string.verification_file_completed)
            }

            OTAState.OTA_STATE_WORKING -> {
                val otaWorking = otaState as OTAWorking
                val message =
                    if (otaWorking.type == JL_Constant.TYPE_CHECK_FILE) getString(R.string.ota_check_file)
                    else getString(R.string.ota_upgrading)
                val progress = Math.round(otaWorking.progress)
                binding.groupUpgrade.show()
                binding.groupScanDeviceLoading.gone()
                binding.tvOtaFileName.apply {
                    if (isAutoTest) {
                        show()
                    } else {
                        gone()
                    }
                }
                binding.tvUpgradeProgress.text = formatProgressText(message, progress)
                binding.pbUpgradeProgress.progress = progress
            }

            OTAState.OTA_STATE_IDLE -> {
                requireActivity().window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
                var isUpgradeSuccess = false
                val otaEnd = otaState as OTAEnd
                when (otaEnd.code) {
                    ErrorCode.ERR_NONE -> {
                        binding.tvUpgradeProgress.text =
                            formatProgressText(getString(R.string.ota_upgrading), 100)
                        binding.pbUpgradeProgress.progress = 100
                        // TODO: 2022/11/1  单次升级结束
                        isUpgradeSuccess = true
                    }

                    ErrorCode.ERR_UNKNOWN -> {}
                    else -> {
                        if (otaEnd.code == ErrorCode.SUB_ERR_OTA_IN_HANDLE) { //正在升级中
                            return
                        } else if (otaEnd.code == ErrorCode.SUB_ERR_DATA_NOT_FOUND) { //未找到升级文件
                            builder.viewModel.readFileList()
                        }
                        val otaMsg = OtaConstant.formatString(
                            "code: %d(0x%X), %s", otaEnd.code, otaEnd.code, otaEnd.message
                        )
                        binding.tvUpgradeResultReason.text = OtaConstant.formatString(
                            "%s\n%s",
                            getString(R.string.ota_reason),
                            otaMsg
                        )
                        isOTAError = true
                    }
                }
                if (!isAutoTest) {//单次升级
                    binding.groupScanDeviceLoading.gone()
                    binding.groupUpgrade.gone()
                    binding.groupUpgradeResult.show()
                    binding.clOtaContent.let {
                        it.layoutParams.let { lp ->
                            if (isUpgradeSuccess) {
                                binding.ivUpgradeResultLogo.setImageResource(R.drawable.ic_success_big)
                                binding.tvUpgradeResultTip.setText(R.string.ota_complete)
                                lp.height = getPixelsFromDp(196)
                            } else {
                                binding.ivUpgradeResultLogo.setImageResource(R.drawable.ic_fail_big)
                                binding.tvUpgradeResultTip.setText(R.string.update_failed)
                                binding.tvUpgradeResultReason.visibility = View.VISIBLE
                                lp.height = FrameLayout.LayoutParams.WRAP_CONTENT
                            }
                            it.layoutParams = lp
                        }
                    }
                }
            }
        }
    }

    private fun formatProgressText(state: String, progress: Int): String =
        OtaConstant.formatString("%s\t\t%d%%", state, progress)

    class Builder(val viewModel: OTAViewModel) : CommonDialog.Builder() {

        init {
            cancelable = false
            widthRate = 1.0f
            gravity = Gravity.BOTTOM
        }

        override fun build(): DialogOTA = DialogOTA(this)

    }
}